### PR TITLE
Issue 610 enhanced border effect in windows

### LINKED ISSUE #610: Make headers separators more visible.
#### Software versions 
Briefcase v1.11.x, Java v1.8.x, operating system: Windows 10

#### Problem description
I think that horizontal and vertical separators of the forms table's header row could be more visible on the Windows system on all Pull, Push and Export tabs. It is not a high priority issue just a small visual improvement. 
Screenshot attached: 
screenshot from 2018-08-21 10-02-24

#### Steps to reproduce the problem
1. Run briefcase on windows os.

### PR DESCRIPTION
**Pull request by huangyz0918**
Closes #610 

#### What has been done to verify that this works as intended?
Enhanced border effect in windows.
#### Why is this the best possible solution? Were any other approaches considered?
Now, It looks like:

image

#### Are there any risks to merging this code? If so, what are they?
No. It only affects Windows.
#### Does this change require updates to documentation? If so, please file an issue at https://github.com/opendatakit/docs/issues/new and include the link below.

### CODE REVIEW

**Comment by dcbriccetti** [comment]
Did we look at ways to solve this that wouldn’t involve any OS-specific code?

**Comment by huangyz0918** [comment]
@dcbriccetti The strange thing is that this method does not work in Windows 10, which looks like:

image

So, this is why I like previous solution. I tried adding a background to avoid using any OS-specific code.
The new commit looks like:

macOS:
image
Windows10:
image

What do you think about this ? @dcbriccetti , or just previous solution?

**Comment by dcbriccetti** [comment]
>this method does not work in Windows 10

By “this method” you mean the approach of increasing the font size? Are you saying it doesn’t increase the font size? My suggestion is that we let the larger font size solve the problem of the headings looking small relative to the rows, and that we don’t worry about background or borders, letting the Look and Feel do what it will.

**Comment by huangyz0918** [comment]
I am sorry that I did not make it clear.
> By “this method” you mean the approach of increasing the font size? Are you saying it doesn’t increase the font size?

The font size was increased in Windows 10, but the border still unclear, so I think is adding a background color will be better. What do you think?

If you think it doesn't matter, I can update this PR as your proposed soon. Thanks.

**Comment by dcbriccetti** [comment]
Thanks for the clarification. I’m concerned about making changes that we don’t fully understand. Or perhaps you do understand them.
Why is this necessary?: `header.setOpaque(false);`
Is it sufficient to only do that, and not change the background color?

If we do go ahead with a table header appearance customization, I suggest you factor it out rather than duplicating the code. You could create a class such as `TableCustomizer` with a static method `customizeHeader` where you put the code, along with comments explaining exactly what we are doing and why. Or you could create an abstract class like `CustomTable` that extends `JTable` and put the code there.

Or perhaps we will decide we don’t understand what’s going on and lack the Swing expertise to do this right, respecting multiple platforms and programmable look and feels (https://docs.oracle.com/javase/tutorial/uiswing/lookandfeel/plaf.html). I don’t know you, @huangyz0918, and perhaps you do deeply understand this.

**Comment by ggalmazor** [comment]
Thanks, @huangyz0918!

I agree with @dcbriccetti that we need to explore the solution a little bit more to be more comfortable before merging. I know this can be frustrating but making this kind of customization work with Swing UIs can be challenging.

If I understood correctly, we are aiming to do the smallest possible customization, to avoid having a snowball effect down the road.

I'd suggest going with the `TableCustomizer`. We are trying to stay away from using inheritance for composition purposes. Having the `ExportFormsTable` and the `TransferFormsTable` *use* a `TableCustomizer` seems like an idea that fits the current design.

**Comment by huangyz0918** [comment]
Thanks for your suggestions! @dcbriccetti @ggalmazor 

> Why is this necessary?: header.setOpaque(false);

In macOS, this may be useless, but it can make the background color available in Windows 10.

> If we do go ahead with a table header appearance customization, I suggest you factor it out rather than duplicating the code.

Agreed. I can create a simple util class (like `TableCustomizer.class`) to hold the custom code.

> we are aiming to do the smallest possible customization.

That's true, since this issue we are talking about is just a tiny improvement. So, I'd like to address it in a easier way.

As for the inconsistent issue with table headers on different platforms (macOS and Windows10), it's because the default `TableCellRenderer` are different in different platforms (this is why I add `header.setOpaque(false);` ), so, what we need to do is set a fixed `TableCellRenderer`.

I propose to add a background color in the table header and make it has the same effect in both macOS and Windows 10.

**Comment by huangyz0918** [comment]
Now, it looks better:
| Windows10 | macOS |
|:---------------:|:-----:|
|windows|macOS|

Please review @dcbriccetti @ggalmazor , thanks.

**Comment by ggalmazor** [comment]
The end result seems OK to me. Here's a screenshot from Ubuntu:
image

File: src/org/opendatakit/briefcase/ui/export/components/ExportFormsTableView.java [hunk 1]
```diff
@@ -50,15 +49,12 @@
 
     addMouseListener(new MouseAdapterBuilder().onClick(this::relayClickToButton).build());
 
-    Dimension formNameDims = getHeaderDimension(HEADERS[FORM_NAME_COL]);
-    Dimension exportStatusDims = getHeaderDimension(HEADERS[EXPORT_STATUS_COL]);
-    Dimension lastExportDims = getHeaderDimension(HEADERS[LAST_EXPORT_COL]);
+    Dimension formNameDims = TableCustomizer.getHeaderDimension(HEADERS[FORM_NAME_COL], this);
```
Comment (ggalmazor) [inline src/org/opendatakit/briefcase/ui/export/components/ExportFormsTableView.java]: I'd suggest adding a static import of all these methods in TableCustomizer to make the code less verbose.

File: src/org/opendatakit/briefcase/ui/export/components/ExportFormsTableView.java [hunk 2]
```diff
@@ -86,14 +82,6 @@
     sorter.sort();
   }
 
-  private Dimension getHeaderDimension(String header) {
```
Comment (ggalmazor) [inline src/org/opendatakit/briefcase/ui/export/components/ExportFormsTableView.java]: Good idea to extract this into the `TableCustomizer` class.

File: src/org/opendatakit/briefcase/ui/reused/TableCustomizer.java [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+import java.awt.Color;
+import java.awt.Dimension;
+import javax.swing.JTable;
+import javax.swing.UIManager;
+import javax.swing.table.JTableHeader;
+
+public class TableCustomizer {
+
+  /**
+   * for header customization.
```
Comment (ggalmazor) [inline src/org/opendatakit/briefcase/ui/reused/TableCustomizer.java]: I'd suggest to change this to something that explains what is this method doing. Something like "Sets properties of the given {@link JTableHeader} to improve its visual appearance across all platforms (Mac, Windows, and Linux)"

File: src/org/opendatakit/briefcase/ui/reused/TableCustomizer.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+   *
+   * @param header
+   */
+  public static void customizeHeader(JTableHeader header) {
+    // to make the background color is available.
+    header.setOpaque(false);
+    header.setBackground(UIManager.getColor("ToolBar.background"));
```
Comment (ggalmazor) [inline src/org/opendatakit/briefcase/ui/reused/TableCustomizer.java]: I'd add a static import for all those `UIManager` methods to reduce code verbosity.

File: src/org/opendatakit/briefcase/ui/reused/TableCustomizer.java [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+import javax.swing.JTable;
+import javax.swing.UIManager;
+import javax.swing.table.JTableHeader;
+
+public class TableCustomizer {
+
+  /**
+   * for header customization.
+   *
+   * @param header
```
Comment (ggalmazor) [inline src/org/opendatakit/briefcase/ui/reused/TableCustomizer.java]: I think these params don't add much value. I'd suggest removing them.

File: src/org/opendatakit/briefcase/ui/reused/TableCustomizer.java [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+   */
+  public static void customizeHeader(JTableHeader header) {
+    // to make the background color is available.
+    header.setOpaque(false);
+    header.setBackground(UIManager.getColor("ToolBar.background"));
+    header.setBorder(UIManager.getBorder("MenuBar.border"));
+  }
+
+  /**
+   * for table customization.
```
Comment (ggalmazor) [inline src/org/opendatakit/briefcase/ui/reused/TableCustomizer.java]: Same comment as with the `customizeHeader(JTableHeader)` method. Put something more descriptive.

File: src/org/opendatakit/briefcase/ui/reused/TableCustomizer.java [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+   *
+   * @param table
```
Comment (ggalmazor) [inline src/org/opendatakit/briefcase/ui/reused/TableCustomizer.java]: Same here, remove it?

File: src/org/opendatakit/briefcase/ui/reused/TableCustomizer.java [hunk 8]
```diff
@@ -0,0 +1,61 @@
...
+   */
+  public static void customizeTable(JTable table) {
+    table.setGridColor(Color.lightGray);
```
Comment (ggalmazor) [inline src/org/opendatakit/briefcase/ui/reused/TableCustomizer.java]: Same: static import `lightGray` from `Color`?

File: src/org/opendatakit/briefcase/ui/reused/TableCustomizer.java [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+    table.setRowHeight(28);
+  }
+
+  /**
+   * to get a dimension from specific header.
```
Comment (ggalmazor) [inline src/org/opendatakit/briefcase/ui/reused/TableCustomizer.java]: Change to "Returns the {@link Dimension} of the given header in the given {@link JTable} using its related cell renderer component's preferred size"

File: src/org/opendatakit/briefcase/ui/reused/TableCustomizer.java [hunk 10]
```diff
@@ -0,0 +1,61 @@
...
+   *
+   * @param table
```
Comment (ggalmazor) [inline src/org/opendatakit/briefcase/ui/reused/TableCustomizer.java]: Same: remove them params?

File: src/org/opendatakit/briefcase/ui/reused/transfer/TransferFormsTableView.java [hunk 11]
```diff
@@ -49,14 +48,11 @@ public TransferFormsTableView(TransferFormsTableViewModel model, String[] header
 
     addMouseListener(new MouseAdapterBuilder().onClick(this::relayClickToButton).build());
 
-    Dimension formNameDims = getHeaderDimension(getHeader(FORM_NAME_COL));
-    Dimension statusDims = getHeaderDimension(getHeader(STATUS_COL));
+    Dimension formNameDims = TableCustomizer.getHeaderDimension(getHeader(FORM_NAME_COL), this);
```
Comment (ggalmazor) [inline src/org/opendatakit/briefcase/ui/reused/transfer/TransferFormsTableView.java]: Same: static import?

File: src/org/opendatakit/briefcase/ui/export/components/ExportFormsTableView.java (hunk 1 above)
Comment (huangyz0918) [reply src/org/opendatakit/briefcase/ui/export/components/ExportFormsTableView.java]: Improved.

File: src/org/opendatakit/briefcase/ui/reused/TableCustomizer.java (hunk 3 above)
Comment (huangyz0918) [reply src/org/opendatakit/briefcase/ui/reused/TableCustomizer.java]: Improved.

File: src/org/opendatakit/briefcase/ui/export/components/ExportFormsTableView.java (hunk 2 above)
Comment (huangyz0918) [reply src/org/opendatakit/briefcase/ui/export/components/ExportFormsTableView.java]: Thank you.

File: src/org/opendatakit/briefcase/ui/reused/TableCustomizer.java [hunk 12]
```diff
@@ -15,43 +15,43 @@
...
   /**
-   * to get a dimension from specific header.
-   *
-   * @param table
-   * @param header
+   * Returns the {@link Dimension} of the given header in the given {@link JTable}
+   * using its related cell renderer component's preferred size.
    */
-  public static Dimension getHeaderDimension(String header, JTable table) {
+  public static Dimension getHeaderDimension(JTable table, String header) {
```
Comment (ggalmazor) [inline src/org/opendatakit/briefcase/ui/reused/TableCustomizer.java]: Nicer arg ordering, thanks!

File: src/org/opendatakit/briefcase/ui/export/components/ExportFormsTableView.java (hunk 2 above)
Comment (dcbriccetti) [reply src/org/opendatakit/briefcase/ui/export/components/ExportFormsTableView.java]: Hey, he was praising *me*. 😀

File: src/org/opendatakit/briefcase/ui/reused/TableCustomizer.java (hunk 9 above)
Comment (dcbriccetti) [reply src/org/opendatakit/briefcase/ui/reused/TableCustomizer.java]: You have mad doc skillz, @ggalmazor.

File: src/org/opendatakit/briefcase/ui/reused/TableCustomizer.java (hunk 10 above)
Comment (dcbriccetti) [reply src/org/opendatakit/briefcase/ui/reused/TableCustomizer.java]: Ha! After I just complimented your English. ;-)

File: src/org/opendatakit/briefcase/ui/reused/TableCustomizer.java (hunk 9 above)
Comment (ggalmazor) [reply src/org/opendatakit/briefcase/ui/reused/TableCustomizer.java]: I don't think the GitHub user @link agrees with you, but ¯\\_(ツ)_/¯

**Comment by dcbriccetti** [comment]
Nice work, @huangyz0918! And good teamwork.

**Pull request merged**